osrf/rmw_zenoh
# Implement instanced message queues with varying depth (and rework rmw_wait) for SERVICES!

This PR is like https://github.com/osrf/rmw_zenoh/pull/27 but for SERVICES (both servers and clients)!

It implements instanced message queues for services (so we no longer need to rely on a static message map.) It also does the same for availability query responses.

This also means that we can now maintain request and response message queues of varying depth!

All other benefits in https://github.com/osrf/rmw_zenoh/pull/27 analogously apply to services. Wait has been updated to reflect this as well.

ALSO: rmw_wait has been edited to ignore guard conditions and (QoS) events. This will suppress the issues we had with invalid QoS warnings, but not solve the underlying issue. (On the other hand, events are supposed to be stubbed out anyway, so...)

## Extra Notes
The bug where a service will not respond to a client that was started before it has been fixed. But the fix is unintuitive and I still do not know why it happens. A very long comment chunk has been added to the fix (which is in `rmw_create_client`) to detail this.

I've also added a magic number to cause clients to not spam service availability checks so often. (I've found that Zenoh queryables sometimes fail or get stuck if they receive queries at more than ~200Hz. Though I did not strictly profile this.)

To test it, first run the client:
```
RMW_IMPLEMENTATION=rmw_zenoh_cpp ros2 run zenoh_ros_services client_main
```

Then run the server in a different process.
```
RMW_IMPLEMENTATION=rmw_zenoh_cpp ros2 run zenoh_ros_services service_main
```

Vice versa should work as before.

Remember to use https://github.com/methylDragon/zenoh_ros_examples/ and remember that you may use --ros-args --log-level debug to see the debug logs.
